File: tests/test_metrics.py
import pytest
import geopandas as gpd
import os
from pathlib import Path

# Create a test file (must start with test_)
# File: test_metrics.py

def test_positive_values():
    values = [1, 2, 3]
    assert all(v > 0 for v in values)

# Example of validating scientific assumptions
def test_no_negative_heights():
    heights = [10.2, 5.4, 3.1]
    assert min(heights) >= 0


def test_load_roads_geojson():
    """Validates that the mountain roads data is present and geographically valid."""
    
    # 1. Get the path of THIS test file (root/scripts/test_file.py)
    this_file = Path(__file__).resolve()
    
    # 2. Go up one level to the root, then into the data folder
    root_dir = this_file.parent.parent
    path = root_dir / "data" / "mount_roads10m.geojson"

    # 3. Check if the file exists
    assert path.exists(), f"Critical Error: {path} is missing!"
    
    # 2. Try loading the file
    gdf = gpd.read_file(path)
    
    # 3. Check for data corruption or empty files
    assert len(gdf) > 0, "The roads GeoJSON loaded but contains zero rows."
    
    # 4. Check for geometry type
    geom_types = gdf.geometry.type.unique()
    
    # Define what we are willing to accept for this workshop
    allowed_types = ['Polygon', 'MultiPolygon', 'LineString', 'MultiLineString']
    
    # Check if the actual types in the file overlap with our allowed list
    assert any(t in allowed_types for t in geom_types), \
        f"Unexpected geometry types found: {geom_types}"

File: tests/test_metrics.R
library(testthat)
library(sf)
library(here)

test_that("addition works correctly", {
  result <- 2 + 3
  expect_equal(result, 5)
})

test_that("mount_roads10m.geojson is present and geographically valid", {
  
  # 1. Check if the file exists (Reproducibility Check)
  # R uses relative paths from the project root/working directory
  path <- here("data", "mount_roads10m.geojson")
  expect_true(file.exists(path), info = paste("Critical Error:", path, "is missing!"))
  
  # 2. Try loading the file
  # st_read is the 'sf' equivalent of gpd.read_file
  # quiet = TRUE prevents the metadata from cluttering your test output
  gdf <- st_read(path, quiet = TRUE)
  
  # 3. Check for data corruption or empty files
  expect_gt(nrow(gdf), 0)
  
  # 4. Check for geometry type
  # st_geometry_type returns a factor; we convert to character to compare
  geom_types <- as.character(unique(st_geometry_type(gdf)))
  
  # Define what we are willing to accept (R types are usually UPPERCASE)
  allowed_types <- c("POLYGON", "MULTIPOLYGON", "LINESTRING", "MULTILINESTRING")
  
  # Check if there is any intersection between found types and allowed types
  expect_true(any(geom_types %in% allowed_types), 
              info = paste("Unexpected geometry types found:", paste(geom_types, collapse = ", ")))
  
  # 5. Scientific Check (Bonus): Ensure a Coordinate Reference System is present
  expect_false(is.na(st_crs(gdf)), info = "Spatial data is missing a CRS!")
})